odl/tomo/operators/ray_trafo.py

```diff
@@ -499,73 +345,51 @@ def __init__(self, range, geometry, **kwargs):
+        # This implementation simply uses a RayTransform to do the job
+        self._ray_trafo = RayTransform(reco_space=range,
+                                       proj_space=domain,
+                                       geometry=geometry,
+                                       **kwargs)
 
-        This method also sets ``self._astra_backprojector`` for
-        ``impl='astra_cuda'`` and enabled cache.
-        """
-        if self.impl.startswith('astra'):
-            backend, data_impl = self.impl.split('_')
-            if data_impl == 'cpu':
-                return astra_cpu_back_projector(
-                    x_real, self.geometry, self.range.real_space, out_real,
-                    **kwargs)
-            elif data_impl == 'cuda':
-                if self._astra_wrapper is None:
-                    astra_wrapper = AstraCudaBackProjectorImpl(
-                        self.geometry, self.range.real_space,
-                        self.domain.real_space)
-                    if self.use_cache:
-                        self._astra_wrapper = astra_wrapper
-                else:
-                    astra_wrapper = self._astra_wrapper
+        # Extra kwargs that can be reused for adjoint etc. These must
+        # be retrieved with `get` instead of `pop` above.
+        self._extra_kwargs = kwargs
 
-                return astra_wrapper.call_backward(x_real, out_real, **kwargs)
-            else:
-                # Should never happen
-                raise RuntimeError('bad `impl` {!r}'.format(self.impl))
+    @property
+    def geometry(self):
+        return self._ray_trafo.geometry
 
-        elif self.impl == 'skimage':
-            return skimage_radon_back_projector(
-                x_real, self.geometry, self.range.real_space, out_real,
-                **kwargs)
-        else:
-            # Should never happen
-            raise RuntimeError('bad `impl` {!r}'.format(self.impl))
+    @property
+    def use_cache(self):
+        return self._ray_trafo.use_cache
+
+    def _call(self, x, out, **kwargs):
+        """Back-projection for the current set-up."""
+
+        return self._ray_trafo.create_impl(self._ray_trafo.use_cache) \
+            .call_backward(x, out, **kwargs)
 
     @property
     def adjoint(self):
         """Adjoint of this operator.
 
+        Returns a copy of the internally used `RayTransform`, thus
+        reusing the cached backend (if `use_cache`).
+
         Returns
         -------
         adjoint : `RayTransform`
         """
-        if self._adjoint is not None:
-            return self._adjoint
-
-        kwargs = self._extra_kwargs.copy()
-        kwargs['range'] = self.domain
-        self._adjoint = RayTransform(self.range, self.geometry,
-                                     impl=self.impl,
-                                     use_cache=self.use_cache,
-                                     **kwargs)
-        return self._adjoint
+        return copy.copy(self._ray_trafo)
```

Comment:
See my suggestion for an inner class. It's largely the same, just without copying (which I don't think is necessary anyway).

Reply:
I like this. Makes `RayTransform` really clean and expressive. I was thinking that perhaps `.adjoint` had to return a new class (as it was before), but I don't think that is sensible with an `@property`.